### DISCUSSION TITLE
Is there a way to reliably programmatically escape/ignore html in mdx?

### DISCUSSION THREAD

**Discussion by Stvad** [mdx-js/mdx#1477] [Q&A] [answered]
**My situation is as follows:**  
I get input content as a set of "blocks" each of the blocks contains a flavored Markdown. A part of this special flavor is that html elements are not rendered. So those block can contain both valid and invalid html (I don't control content).  

What I want to do is to assemble a set of blocks into a pages rendered by MDX. Which works well most of the time, but things fail parsing (and the build) when there is malformed html present in the source.   

My question is - is there a reliable programmatic way to escape html within mdx so it will get parsed as a plain text. Or configure parser to ignore html elements in particular contexts somehow (by wrapping them in special element or something like that)?

https://maxschmitt.me/posts/escape-html-mdx/ is the only thing I found on the topic and it doesn't really provide a solution that works in all contexts (inline, multiline, and not disfiguring code inside code-fences)

**Context:** Building a static site with `gatsby-plugin-mdx`

**Comment by ChristianMurphy** [comment, marked as answer]
HTML and JSX are incompatible, `mdx` parses JSX.

There are a few options:
1. validate HTML tags to avoid invalid content
2. use https://github.com/remarkjs/remark#readme to parse HTML (`mdx` builds on `remark`, `mdx` replaces `remark`'s HTML parser with a JSX parser)

**Reply by Stvad** [reply]
I don't want to parse HTML - I want to reliably ignore it & display it as plaintext (though if it's parsed on a "best effort" basis without breaking the build I won't mind). My core requirement is for build not to break (https://github.com/gatsbyjs/gatsby/discussions/30067)

>validate HTML tags to avoid invalid content 

I want to display content regardless, it's plain markdown (with no html) in vast majority of the times.  
I suppose if I somehow validate "can mdx parse this block" - I can replace specific problematic blocks with error message.

**Reply by ChristianMurphy** [reply]
See https://github.com/mdx-js/mdx/discussions/1236#discussioncomment-57133 in particular the third point.
MDX parses JSX to spec, which means that just like in JSX invalid HTML/JSX will throw an error, this can be handled with a React error boundary to gracefully fallback, but there is not partial parsing in JSX.

**Reply by Stvad** [reply]
Thanks! Error boundary approach seems promising, let me experiment with that!

**Reply by Stvad** [reply]
Hmm, it seems React Error boundary operates at wrong level to help with this. It can handle runtime errors but not syntax ones. Unless I use MDX Runtime, but that is not recommended for production as I understand it?  

Gatsby build fails as follows:

```
 ERROR #11321  PLUGIN

"gatsby-plugin-mdx" threw an error while running the onCreateNode lifecycle:

unknown: Expected corresponding JSX closing tag for <b> (519:2)

  517 |   <b> malformed
  518 |
> 519 |   </ErrorBoundary>
      |   ^
  520 |   </div>
  521 |
  522 | </div>undefined: unknown: Expected corresponding JSX closing tag for <b> (519:2)

  517 |   <b> malformed
  518 |
> 519 |   </ErrorBoundary>
      |   ^
  520 |   </div>
  521 |
  522 | </div>

```

**Reply by Stvad** [reply]
One thing I can try is to do a kind of build-time error boundary: 
- try to run mdx compiler for the block https://mdxjs.com/advanced/api#compiler 
- If that fails replace block with "there were an error here" 
- if parsing is successful - throw results away & let gatsby-plugin-mdx do the final rendering of the original content. 

Will try that, but it'd mean doing the parsing twice, which is bound to have perf impact. Please let me know if you know of a better approach here!

**Reply by ChristianMurphy** [reply]
I'm not as personally familiar with the Gatsby plugin.
That may work.

I know for webpack in MDX
https://github.com/mdx-js/mdx/blob/c2134e15089c9f90140c95b4556c334dbc4f118b/packages/loader/index.js#L1-L27
and in XDM
https://github.com/wooorm/xdm/blob/978f5ee081664a57cb3d6246f01e93831e3cc331/lib/integration/webpack.js#L1-L21

the code is light enough, that swapping it for a custom loader that drops in a fallback component when there is an error wouldn't be too complex, and would avoid the double parse perf overhead.

Something similar may be possible with the Gatsby plugin.

**Reply by Stvad** [reply]
Thanks, will explore that too. For now - did an experiment doing what I described above and it works. For posterity it looks ~ like:

```typescript
function canParse(mdStr: string) {
  try {
    mdx.sync(mdStr)
    return true
  } catch (e) {
    console.log("problematic block:", mdStr)
    console.log(e)
    return false
  }

}

// to be used like
<div class="block-content">

  ${canParse(mdStr) ? mdStr : `<>{\`${mdStr}⚠️\`}</>`}
  
  </div>

```

**Reply by Stvad** [reply]
I appreciate you help on this!   
I'm not quite sure if it's correct to mark this as answered, as although given approach works for my situation, this doesn't actually answer the question about escaping html 🤔

**Reply by ChristianMurphy** [reply]
It has, repeatly, the answer is no, it is not supported, and would not be supported without a change to the language and parser.
MDX parses as markdown or JSX, if the JSX is invalid it will error.
Unclosed and incomplete is invalid JSX.

Changing this behavior would mean changing the language.
You are welcome to open a request for comments to show how JSX and HTML would be differentiated https://github.com/mdx-js/specification#mdx
But be aware that it is a broader and more complex discussion than "escape the invalid HTML", there is no HTML in MDX, only markdown and JSX.

**Reply by Stvad** [reply]
Hmm, to double check - do I understand correctly that you've interpreted my questions as "can I escape what seems like invalid html exclusively while leaving JSX intact"?  

Because I'm ok escaping JSX/any taglike object in a given string too.   
In fact partial approaches described in https://maxschmitt.me/posts/escape-html-mdx/ almost work.  
They are quite ad-hoc but afaik doing something like `string.replace(/</g, "{`<`}")` mostly works, (but doesn't quite do what you want inside code-fences, which can probably be fixed with more checks for that).  What I was wondering about if there was a better way 🤷‍♂️.  

And I think on balance doing double-parsing approach above is probably a better solution for my use-case (though I had to handle some  unexpected challenges there too e.g. https://github.com/mdx-js/mdx/discussions/1485) 

Thanks again for all your help!

**Reply by ChristianMurphy** [reply]
I interpret is as you want to show anything other than the raw text with MDX.
And the answer is still no, because JSX and and markdown can be mixed in MDX, the markdown surrounding also becomes unparseable.

Alternatively you could parse it as plain markdown or don't parse it all, show the raw text.

**Reply by ChristianMurphy** [reply]
Put another way, MDX in inherently Markdown + JSX.
Remove either piece and that thing ceases to be MDX.
If you want to parse just Markdown (remark), Markdown and HTML (remark + rehype), or just JSX (estree), there are tools which can help.
But MDX is not that tool.

**Reply by ChristianMurphy** [reply]
Taking a step back, there may be a deeper issue here, and that question is: why is there invalid content being built?
If the content is being run through a CMS/authoring system, the content can be validated on the fly, or if the authoring experience is out of your control, when it is being ingested into Git or into a CMS it can be validated at that time.

---

Remember MDX is Markdown + JSX. Take Markdown out of the equation for a moment.
If a person commits a JSX file with a missing bracket that crashes the babel build in your Gatsby project, what would you do?
Open an issue with babel asking "how to programmatically ignore invalid JavaScript?" Or ask the person who committed the file to please add the missing bracket?

**Reply by Stvad** [reply]
## Background
I'd like to add a bit more of a background to hopefully explain why my requirements make sense (they do at least from the inside 😉) 

I work on a project to make it easy to publish notes you've made inside https://roamresearch.com/ (can recommend checking it out if you haven't seen it before) as a static site (digital garden) - https://roam.garden/
So people *originally write content in a different tool and mostly without concern for using it with my publishing service*.
Then they export their content and import it to my service, at which point I try to build a static website from it. 
### This has the following implications:
  - I can't validate content before the build time
  - It'd be really bad customer experience for me to reject an input that is considered valid in the original system (or crash during the build which `gatsby-plugin-mdx` tends to do any time it catches a whiff of syntactially invalid mdx)
### So my goals here are as follows: 
  - Never ever break the build
  - For things that can't be properly fully parsed generate the best representation I can
### How does data look and what I do with it
Fortunately the structure of the data I get is reasonably amenable to the second goal stated above. I get the notes in a JSON that represents a hierarchy of blocks, and each of the blocks is a semi-independent specially flavoured markdown document.
_A simplified page example:_
```json
{
  "title": "test",
  "children": [
    {
      "string": "> I'm markdown",
      "children": [
        {
          "string": "*nested hierarchically*"
        }
      ]
    },
    {
      "string": "and I can have all kinds of weird things",
      "children": [
        {
          "string": "<%J:\n```javascript\nlet reg_parent_string = new RegExp('```clojure\\s(.+)```','s');\n```%>"
        }
      ]
    }
  ]
}
 ```

From this I'm going to generate an MDX document which would look something like the following:
```mdx
# test

// there is no indents IRL as it interferes with parsing             
<Block content={
    <>

        > I'm markdown

    </>
}>
    <Block content={
        <>

            *nested hierarchically*

        </>
    }>

    </Block>
</Block>

<Block content={
    <>

        and I can have all kinds of weird things

    </>
}>
    <Block content={
        <>
            {/* An example of graceful fallback accomplished with double parsing strategy */}             
            <BlockContentFailedParsing content={`<%J:
\`\`\`javascript
let reg_parent_string = new RegExp('\`\`\`clojure\s(.+)\`\`\`','s');
\`\`\`%>`}/>

        </>
    }>
    </Block>
</Block>
```
See an **example of the full export** of the test page here: [canary.txt](https://github.com/mdx-js/mdx/files/6176905/canary.txt) and it in the rendered form https://vlad.roam.garden/?stackedPages=Garden-Lorem-Ipsum

As shown above - Blocks can have all kinds of weird things in them, but fortunately I can treat each block semi-independently and if there is something bad in a given block - I can gracefully and gradually degrade the rendering behavior _just for that block_ while rendering other things on the page properly. 
Which I currently do via [double parsing approach](https://github.com/mdx-js/mdx/discussions/1477#discussioncomment-452585) discussed above.

The **overall document is MDX** but each individual block is **Markdown** (though I use MDX functionality like supplying custom components to override standard ones for them)

To achieve the stated graceful degradation goals - each given block can be treated as follows from the parsing standpoint (in the order of preference):
  - Parse it as full MDX if possible 
  - If can't parse it as MDX because JSX parts seem invalid - escape anything that looks like JSX and try to parse it as Markdown (Original subject of this question, currently not done)
  - If all attempts at parsing failed - display block as plain text (currently implemented with double parsing)

This degradation behavior is implemented at "code generation" time currently when I assemble the full MDX document that would later be parsed by gatsby-plugin-mdx

I hope this helps explain what and why I'm trying to do 🙂.

### Potential alternative approaches
One thing I was wondering from your messages above whether it's possible to parse the internal content of the blocks as pure Markdown but then plug the results of that into the larger document/use it with MDXProvider somehow?
It sounds complicated and I'm not sure it fully possible or better then current approach 🤷‍♂️

**Reply by wooorm** [reply]
I don’t see MDX coming out of the API. Why are you trying to parse it as such?

**Reply by Stvad** [reply]
The overall document/page I generate is MDX as I'd like to use custom components and such and that is what finally gets transformed into a Gatsby node.
 
As noted above, each individual block is **Markdown**, but I want to use MDX functionality like supplying custom components to override standard ones for them. 
So I'm using the functionality discussed in #628  to assemble MDX document with Markdown interleaved.

As I say in the "Potential alternative approaches" section - If I can assemble pipeline that would parse the Markdown blocks first independently (but allowing me to use the gatsby-remark-*/etc plugins) & then allow me to assemble it into the larger document & and use MDXProvider with all that - it can be an alternative. I'm not really sure how to do that and it seems like a more complicated approach from the outset vs the current one.   
If you have any pointers in this direction - I'd love to explore them though!

**Reply by wooorm** [reply]
> So I'm using the functionality discussed in #\628 to assemble MDX document with Markdown interleaved.

That issue is about authoring MDX. Here you’re asking about embedded markdown inside MDX, where the rules of MDX do not apply.

>  If I can assemble pipeline that would parse the Markdown blocks first independently […] & then allow me to assemble it into the larger document & and use MDXProvider with all that

☝️ Yep, this is what you’re looking for. You’ve found Y!

> (but allowing me to use the gatsby-remark-*/etc plugins)

Can you elaborate what you mean by this?

> I'm not really sure how to do that and it seems like a more complicated approach from the outset vs the current one.

It’s a bit more involved, but the problem with your current approach is that it can’t work.

**Reply by Stvad** [reply]
>☝️ Yep, this is what you’re looking for. You’ve found Y!

Maybe, it feels more complicated than X though 😛, and X works. Y has potential to be more reliable though, so I'd really appreciate pointers on how one might go about setting up such a pipeline!

> Can you elaborate what you mean by this?

Current final parsing & transformation is currently performed by `gatsby-plugin-mdx` and uses both raw remark-*, rehype-* plugins and `gatsby-remark-*` plugins (e.g. `gatsby-remark-double-parenthesis-link` or `gatsby-remark-images`). The hypothetical alternative mechanism should be able to do the same.  

And then presumably allow me to assemble the final page in MDX (or raw JSX?) so I can use custom components and override the standard markdown components...

**Comment by EryouHao** [comment]
@Stvad Have you solved this problem, please? Do you have any solutions?

**Reply by wooorm** [reply]
There is no problem: html does not exist in mdx. JSX exists. But, the above question has a lengthy discussion. Your "is there a solution" does not add anything new to warrant a different answer.

**Reply by Stvad** [reply]
something like https://github.com/orgs/mdx-js/discussions/1477#discussioncomment-452585 is the thing I stuck with for my use-case. It's not ideal but kind-of works
